pkg/util/active_groups.go

```diff
@@ -0,0 +1,156 @@
+	totalGroups := len(ag.timestampsPerUser[userID])
+	if totalGroups == 0 {
+		ag.mu.RUnlock()
+		return nil
+	}
+
+	inactiveGroups := make([]string, 0, totalGroups)
+	groupTimestamps := ag.timestampsPerUser[userID]
+
+	for group, ts := range groupTimestamps {
+		if ts.Load() <= deadline {
+			inactiveGroups = append(inactiveGroups, group)
+		}
+	}
+	ag.mu.RUnlock()
+
+	if groupTimestamps == nil || len(inactiveGroups) == 0 {
+		return nil
+	}
+
+	// Cleanup inactive groups
+	deletedGroups := make([]string, 0, len(inactiveGroups))
+	ag.mu.Lock()
+	defer ag.mu.Unlock()
+
+	for _, inactiveGroup := range inactiveGroups {
+		groupTs := ag.timestampsPerUser[userID][inactiveGroup]
+		if groupTs != nil && groupTs.Load() <= deadline {
+			delete(ag.timestampsPerUser[userID], inactiveGroup)
+			deletedGroups = append(deletedGroups, inactiveGroup)
+		}
+	}
+
+	return deletedGroups
+}
+
+type ActiveGroupsCleanupService struct {
+	services.Service
+
+	activeGroups    *ActiveGroups
+	cleanupFuncs    []func(string, string) // Takes userID, group
+	inactiveTimeout time.Duration
+}
+
+type UserGroupMetricsCleaner interface {
+	RemoveGroupMetricsForUser(userID, group string)
+}
+
+func NewActiveGroupsCleanupService(cleanupInterval, inactiveTimeout time.Duration, maxGroupsPerUser int, cleanupFns ...func(string, string)) *ActiveGroupsCleanupService {
+	s := &ActiveGroupsCleanupService{
+		activeGroups:    NewActiveGroups(maxGroupsPerUser),
+		cleanupFuncs:    cleanupFns,
+		inactiveTimeout: inactiveTimeout,
+	}
+
+	s.Service = services.NewTimerService(cleanupInterval, nil, s.iteration, nil).WithName("active groups cleanup")
+	return s
+}
+
+func NewActiveGroupsCleanupWithDefaultValues(maxGroupsPerUser int, cleanupFns ...func(string, string)) *ActiveGroupsCleanupService {
+	return NewActiveGroupsCleanupService(1*time.Minute, 20*time.Minute, maxGroupsPerUser, cleanupFns...)
+}
+
+func (ag *ActiveGroups) ActiveGroupLimitExceeded(userID, group string) bool {
+	ag.mu.RLock()
+	defer ag.mu.RUnlock()
+
+	_, containsGroup := ag.timestampsPerUser[userID][group]
+	return !containsGroup && len(ag.timestampsPerUser[userID]) >= ag.maxGroupsPerUser
+}
+
+func (s *ActiveGroupsCleanupService) UpdateActiveGroupTimestamp(user, group string, now time.Time) string {
+	// Does not track empty label
+	if group == "" {
+		return group
+	}
+
+	if s.activeGroups.ActiveGroupLimitExceeded(user, group) {
+		group = "other"
+	}
+	s.activeGroups.UpdateGroupTimestampForUser(user, group, now)
+	return group
+}
+
+func (s *ActiveGroupsCleanupService) iteration(_ context.Context) error {
+	for userID := range s.activeGroups.timestampsPerUser {
+		inactiveGroups := s.activeGroups.PurgeInactiveGroupsForUser(userID, time.Now().Add(-s.inactiveTimeout).UnixNano())
+		for _, group := range inactiveGroups {
+			for _, cleanupFn := range s.cleanupFuncs {
+				cleanupFn(userID, group)
+			}
+		}
+	}
+	return nil
+}
+
+func (s *ActiveGroupsCleanupService) Register(metricsCleaner UserGroupMetricsCleaner) {
```

Comment:
Please comment what it does and in particular that it's **not** concurrency safe (it's OK, but better to comment given the rest of this struct is concurrency safe).